materialize-redshift/driver.go

```diff
@@ -877,64 +877,136 @@ func (d *transactor) commit(ctx context.Context, fenceUpdate string, hasUpdates
 				return fmt.Errorf("creating store table: %w", err)
 			}
-			batch.Queue(b.storeUpdateDeleteExistingSQL)
-			batch.Queue(b.storeUpdateSQL)
-		}
-	}
 
-	// The fence update is always the last query in the batch.
-	batch.Queue(fenceUpdate)
-
-	res := txn.SendBatch(ctx, &batch)
-
-	for idx := 0; idx < batch.Len()-1; idx++ {
-		if _, err := res.Exec(); err != nil {
-			return fmt.Errorf("exec store batch index %d: %w", idx, err)
+			if copy, err := copySql(fmt.Sprintf("flow_temp_table_%d", idx)); err != nil {
+				return err
+			} else if _, err := txn.Exec(ctx, copy); err != nil {
+				return handleCopyIntoErr(ctx, txn, objectLocation, b.target.Identifier, err)
+			} else if _, err := txn.Exec(ctx, b.mergeIntoSQL); err != nil {
+				return fmt.Errorf("merging to table '%s': %w", b.target.Identifier, err)
+			}
+		} else {
+			// Can copy directly into the target table since all values are new.
+			if copy, err := copySql(b.target.Identifier); err != nil {
+				return err
+			} else if _, err := txn.Exec(ctx, copy); err != nil {
+				return handleCopyIntoErr(ctx, txn, objectLocation, b.target.Identifier, err)
+			}
 		}
 	}
 
-	fenceRes, err := res.Exec()
-	if err != nil {
+	if fenceRes, err := txn.Exec(ctx, fenceUpdate); err != nil {
 		return fmt.Errorf("fetching fence update rows: %w", err)
 	} else if fenceRes.RowsAffected() != 1 {
 		return errors.New("this instance was fenced off by another")
-	}
-	if err := res.Close(); err != nil {
-		return fmt.Errorf("closing store batch: %w", err)
-	}
-
-	if err := txn.Commit(ctx); err != nil {
+	} else if err := txn.Commit(ctx); err != nil {
 		return fmt.Errorf("committing store transaction: %w", err)
 	}
 
 	return nil
 }
 
+// handleCopyIntoErr queries the `sys_load_error_detail` table for relevant COPY INTO error details
+// and returns a more useful error than the opaque error returned by Redshift. This function will
+// always return an error. `sys_load_error_detail` is queried instead of `stl_load_errors` since it
+// is available to both serverless and provisioned versions of Redshift, whereas `stl_load_errors`
+// is only available on provisioned Redshift.
+func handleCopyIntoErr(ctx context.Context, txn pgx.Tx, objectLocation string, table string, copyIntoErr error) error {
+	// The transaction has failed. It must be finish being rolled back before using its underlying
+	// connection again.
+	txn.Rollback(ctx)
+	conn := txn.Conn()
+
+	log.WithFields(log.Fields{
+		"table": table,
+		"err":   copyIntoErr.Error(),
+	}).Warn("COPY INTO error")
+
+	loadErrInfo, err := getLoadErrorInfo(ctx, conn, objectLocation)
+	if err != nil {
+		return fmt.Errorf("COPY INTO error for table '%s' but could not query sys_load_error_detail: %w", table, err)
+	}
+
+	log.WithFields(log.Fields{
+		"errMsg":    loadErrInfo.errMsg,
+		"errCode":   loadErrInfo.errCode,
+		"colName":   loadErrInfo.colName,
+		"colType":   loadErrInfo.colType,
+		"colLength": loadErrInfo.colLength,
+	}).Warn("loadErrInfo")
```

Comment:
This seems redundant given that we have a catch-all case below, or is it used for some other purpose?

Reply:
Hopefully it is redundant all the time, I have it here just in case there's something unexpected that happens with the explicit error codes handled in the switch statement. For example, if a `1216` code actually ends up being something other than an "Input line is not valid", that way we'll have all of the information in the logs at least.